File: lib/fulltext_reader.rb
# See README for Copyright and License information

require 'chunked_io'

module FateSearch
  class FulltextReader

    def initialize(options = {})
      raise ArgumentError, "Need the path to the suffix array file" unless options[:path]
      @io = ChunkedIo.open(options[:path], "rb")
    end

    def get_data(offset, size)
      @io.pos = offset[0]
      @io.read(size)
    end
          
    # Gets the record start position and the size and reads the block
    def offset_to_record_data(offset)
      record_start, size = offset_to_record_start(offset)
      get_data(record_start, size)
    end  
    
    def hit_to_record_data(hit)
      @io.pos = hit.record_offset
      size = @io.read(4).unpack("V")[0]
      @io.pos = hit.record_offset
      @io.read(size)
    end  
    
    def get_primary_key(data_block)
      # Primary key is the second offset in the header, stored as a long
      primary_key = data_block[4..7].unpack("V")[0]
    end

    def get_fields(data_block)
      # Header: 8 == total field data size, primary key
      header_size = 8
      footer_size = 5
      size = data_block.size
      fields = []
      field_start = header_size
      while field_start <= size - (header_size + footer_size)
        # Fields are structured: data size, data, \0
        field_size = data_block[(field_start)..(field_start+3)].unpack("V")[0] 
        fields << data_block[(field_start+4)..(field_start+4+field_size-1)]
        field_start += field_size + 5 
      end
      fields
    end
        
    def rank_offsets(hits, weights, term, limit = 10, compare_size = false)
      scores = Hash.new{|h,k| h[k] = 0.0}
      current = hits.size
      size = term.size.to_f
      percent_diff = 1
      hits.each {|hit|
        # TODO, could maybe add one more number, the suffix size, into the array
        if (compare_size)
          @io.pos = hit.offset
          true while @io.getc != 0
          text_size = @io.pos - hit.offset
          diff = (size - text_size).abs
          percent_diff = 1 - (diff / size)
        end  
        scores[hit.record_offset] += (weights[hit.field_id] * percent_diff) + current
        current -= 1
      }
      sorted_offsets = scores.sort_by{|offset,score| -score}
      # Return the results as data blocks
      blocks = []
      0.upto(limit-1) {|index|
        break if index >= sorted_offsets.size
        offset, score = sorted_offsets[index]
        @io.pos = offset
        size = @io.read(4).unpack("V")[0]
        @io.pos = offset
        data = @io.read(size)        
        blocks << [get_primary_key(data), get_fields(data), score]
      }  
      blocks
    end

  private
    # TODO this could be replaced with a single seek to the index in a record map
    # Instead of passing offset, the suffix index would be compared to the map
    # A simpler solution is to double the size of the suffix array and store the
    # record start (and possily a single byte for field index?) It is likely there
    # is a better construction for a number relative to the suffix offset so it 
    # would not need to be a four-byte long
    def offset_to_record_start(offset)
      # Read to the end of the data (from the offset) which is marked by a footer
      # Footer: \0
      @io.pos = offset      
      # Which is faster? this or @io.gets("\0")
      true while @io.getc != 0
      
      # Repeatedly loop through the fields to the record footer
      while true       
        # Read the size, it may be the record size, or subsequent field
        size = @io.read(4).unpack("V")[0]
        c = @io.getc
        # If the size is followed by a null then it is the record footer 
        # Unless the size is 0, then it is just an empty field
        if (c == 0 && size != 0)          
          end_pos = @io.pos
          record_start = (end_pos-size)          
          break
        else
          # Not a null, skip the field data (and trailing null)
          @io.pos = @io.pos + size
        end
      end        
      [record_start, size]
    end  
  end
end

File: javascripts/fulltext_reader.js
  /**
 */
function FulltextReader(baseURL){
  var _reader = new ChunkedFileReader(baseURL);

  this.getData = function(offset, size) {
    _reader.seek(offset);
    var block = _reader.read(size);
    return _reader.blockToString(block);
  }
  
  this.hitToRecordData = function(hit) {
    _reader.seek(hit["recordOffset"]);
    var block = _reader.read(4);
    var size = _reader.blockToNumber(block);
    _reader.seek(hit["recordOffset"]);
    return _reader.read(size);    
  }
  
  this.getPrimaryKey = function(block) {
    // Primary key is the second offset in the header, stored as a long
    return _reader.blockToNumber(block.slice(4,7));    
  }
  
  this.getFields = function(block) {
    // Header: 8 == total field data size, primary key
    var headerSize = 8;
    var footerSize = 5;
    var size = block.length;
    var index = 0;
    var fields = [];
    var fieldStart = headerSize;
    while (fieldStart <= size - (headerSize + footerSize)) {
      // Fields are structured: data size, data, \0
      fieldSize = _reader.blockToNumber(block.slice(fieldStart,(fieldStart+3)));
      fields[index++] = _reader.blockToString(block.slice((fieldStart+4),(fieldStart+4+fieldSize)));
      fieldStart += fieldSize + 5;
    }
    return fields;      
  }  
  
  this.rankOffsets = function(term, hits, weights, compareSize, page, perPage) {
    page = page || 1;
    perPage = perPage || 10;
    compareSize = compareSize || false;
    var scores = [];
    var current = hits.size();
    var size = term.length;
    var percentDiff = 1;
    while (current > 0) {
      var hit = hits.get(current-1);
      // Comparing the size of hit to term length gives the importance of the 
      // term to the field. In general this can be approximated just as well 
      // based on average length of the field. If you compare, it requires 
      // reading every hit and looking up a lot of shards.
      //
      // For example: if you search for "John" in the sample corpus, the 
      // highest ranked item is "Johnson Johnson." If you compare size though,
      // "Johns John" is the highest ranked item. 
      //
      // Eventually we may just store the fieldSize in with the suffixes.
      if (compareSize) {
        _reader.seek(hit["offset"]);
        while (true) { if (_reader.getc() == 0) break; }
        var textSize = _reader.pos() - hit["offset"];
        var diff = Math.abs(size - textSize);
        percentDiff = 1 - (diff / size);
      }
      // Each item in the scores array will have [offset, score]
      scores[hit["recordOffset"]] = scores[hit["recordOffset"]] || [hit["recordOffset"], 0];
      scores[hit["recordOffset"]][1] += (weights[hit["fieldId"]] * percentDiff) + current;
      current -= 1;
    }
    // Compact the array
    var offsets = [];
    for (var i=0; i<scores.length; i++) {
      if (scores[i]) offsets.push(scores[i]); 
    }
    // Sort the array by score (descending)
    offsets.sort(function(a,b){ return b[1] - a[1] });
    var blocks = []
    var start = ((page-1)*perPage);
    for (var i=start; i<start+perPage; i++) {
      if (i > offsets.length - 1) break;
      item = offsets[i];
      _reader.seek(item[0]);
      var size = _reader.blockToNumber(_reader.read(4));
      _reader.seek(item[0]);
      var block = _reader.read(size);
      blocks.push([this.getPrimaryKey(block), this.getFields(block), item[1]]);
    }
    return blocks;    
  }
}